src/index.rs

```diff
@@ -574,55 +554,66 @@ impl ResourceIndex {
     }
 }
 
-fn discover_paths<P: AsRef<Path>>(
-    root_path: P,
-) -> HashMap<CanonicalPathBuf, DirEntry> {
+/// Discovers all files under the specified root path
+///
+/// Returns a hashmap of canonical file paths to directory entries
+fn discover_files<P: AsRef<Path>>(root_path: P) -> HashMap<PathBuf, DirEntry> {
     log::debug!(
         "Discovering all files under path {}",
         root_path.as_ref().display()
     );
 
-    WalkDir::new(root_path)
+    let mut discovered_files = HashMap::new();
+    let walker = WalkDir::new(root_path)
         .into_iter()
-        .filter_entry(|entry| !is_hidden(entry))
-        .filter_map(|result| match result {
+        .filter_entry(|entry| {
+            // skip hidden files and directories
+            !entry
+                .file_name()
+                .to_string_lossy()
+                .starts_with('.')
+        });
+
+    for entry in walker {
+        match entry {
             Ok(entry) => {
-                let path = entry.path();
+                let path = entry.path().to_path_buf();
                 if !entry.file_type().is_dir() {
-                    match CanonicalPathBuf::canonicalize(path) {
-                        Ok(canonical_path) => Some((canonical_path, entry)),
+                    // canonicalize the path to avoid duplicates
+                    match fs::canonicalize(&path) {
+                        Ok(canonical_path) => {
+                            discovered_files.insert(canonical_path, entry);
```

Comment:
After canonicalization, we should trim the common prefix ("root").

Reply:
That's related to #86. Do you want me to implement it in this PR?

Reply:
Ah yeah, we can do it in separate PR.

Reply:
After spending some time modifying this PR based on #72, I've realized that it might not be a good idea to combine all the changes from both PRs into a single PR. #72 has significant API changes for `ResourceIndex`, and I think we should discuss these changes further in #91 before proceeding.

This PR focuses only on refactoring certain parts of the code to enhance readability. It doesn't alter any logic for `arklib` so it might be a good idea to keep it simple.
What do you think?